packages/node-core/src/api.service.ts

```diff
@@ -58,6 +62,104 @@ export abstract class ApiService<A = any, SA = any, B = any> implements IApi<A,
     const apiInstance = this.connectionPoolService.api;
     return apiInstance.unsafeApi;
   }
+  async createConnections(
+    network: ProjectNetworkConfig & {chainId: string},
+    createConnection: (endpoint: string) => Promise<IApiConnectionSpecific>,
+    getChainId: (connection: IApiConnectionSpecific) => Promise<string>,
+    postConnectedHook?: (connection: IApiConnectionSpecific, endpoint: string, index: number) => void
+  ): Promise<void> {
+    const endpointToApiIndex: Record<string, IApiConnectionSpecific> = {};
+
+    const failedConnections: Map<number, string> = new Map();
+
+    for await (const [i, endpoint] of (network.endpoint as string[]).entries()) {
+      try {
+        const connection = await createConnection(endpoint);
+        this.eventEmitter.emit(IndexerEvent.ApiConnected, {
+          value: 1,
+          apiIndex: i,
+          endpoint: endpoint,
+        });
+
+        if (postConnectedHook) {
+          postConnectedHook(connection, endpoint, i);
+        }
+
+        if (!this.networkMeta) {
+          this.networkMeta = connection.networkMeta;
+        }
+
+        const chainId = await getChainId(connection);
+
+        if (network.chainId !== chainId) {
+          throw this.metadataMismatchError('ChainId', network.chainId, chainId);
+        }
+
+        endpointToApiIndex[endpoint] = connection;
+      } catch (e) {
+        logger.error(`Failed to init ${endpoint}: ${e}`);
+        endpointToApiIndex[endpoint] = null as unknown as IApiConnectionSpecific;
+        failedConnections.set(i, endpoint);
+      }
+    }
+
+    if (Object.values(endpointToApiIndex).every((value) => value === null)) {
+      throw new Error('All endpoints failed to initialize. Please add healthier endpoints');
+    }
+
+    await this.connectionPoolService.addBatchToConnections(endpointToApiIndex);
+
+    // Retry failed connections in the background
+    for (const [index, endpoint] of failedConnections) {
+      this.retryConnection(createConnection, getChainId, network, index, endpoint, 0, postConnectedHook);
+    }
+  }
+
+  retryConnection(
+    createConnection: (endpoint: string) => Promise<IApiConnectionSpecific>,
+    getChainId: (connection: IApiConnectionSpecific) => Promise<string>,
+    network: ProjectNetworkConfig & {chainId: string},
+    index: number,
+    endpoint: string,
+    attempt: number,
+    postConnectedHook?: (connection: IApiConnectionSpecific, endpoint: string, index: number) => void
+  ): void {
+    if (attempt < 5) {
+      //eslint-disable-next-line @typescript-eslint/no-misused-promises
+      setTimeout(async () => {
```

Comment:
It might be worth having a backoff function as we use the same concept a few places now. Also there needs to be a reference to this timeout